examples/cpp_api/png_ingestion_webp.cc

```diff
@@ -0,0 +1,369 @@
+#include <cstdio>
+
+#include <png.h>
+// Note: on some macOS platforms with a brew-installed libpng, use this instead:
+//#include <libpng16/png.h>
+
+// Include the TileDB C++ API headers
+#include <tiledb/tiledb>
+
+using namespace tiledb;
+
+std::vector<uint8_t> read, write;
+
+// Colorspace stride
+const unsigned colorspace = TILEDB_WEBP_RGBA;
+const unsigned pixel_depth = colorspace < TILEDB_WEBP_RGBA ? 3 : 4;
+const float quality_factor = 100.0f;
+const bool lossless = true;
+
+/**
+ * Reads a .png file at the given path and returns a vector of pointers to
+ * the pixel data in each row. The caller must free the row pointers.
+ *
+ * This is a modified version of: https://gist.github.com/niw/5963798
+ * "How to read and write PNG file using libpng"
+ * (C) 2002-2010 Guillaume Cottenceau
+ * Redistributed under the X11 license.
+ */
+std::vector<uint8_t*> read_png(
+    const std::string& path, unsigned* width, unsigned* height) {
+  std::vector<uint8_t*> row_pointers;
+
+  // Get the image info.
+  auto fp = fopen(path.c_str(), "rb");
+  png_structp png =
+      png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+  png_infop info = png_create_info_struct(png);
+  setjmp(png_jmpbuf(png));
+  png_init_io(png, fp);
+  png_read_info(png, info);
+
+  *width = png_get_image_width(png, info);
+  *height = png_get_image_height(png, info);
+  uint8_t color_type = png_get_color_type(png, info),
+          bit_depth = png_get_bit_depth(png, info);
+
+  // Read any color_type into 8bit depth, RGBA format.
+  // See http://www.libpng.org/pub/png/libpng-manual.txt
+  if (bit_depth == 16)
+    png_set_strip_16(png);
+
+  if (color_type == PNG_COLOR_TYPE_PALETTE)
+    png_set_palette_to_rgb(png);
+
+  // PNG_COLOR_TYPE_GRAY_ALPHA is always 8 or 16bit depth.
+  if (color_type == PNG_COLOR_TYPE_GRAY && bit_depth < 8)
+    png_set_expand_gray_1_2_4_to_8(png);
+
+  if (png_get_valid(png, info, PNG_INFO_tRNS))
+    png_set_tRNS_to_alpha(png);
+
+  // If example is ran with RGBA ensure alpha is present
+  if (colorspace == TILEDB_WEBP_RGBA || colorspace == TILEDB_WEBP_BGRA) {
+    // These color_type don't have an alpha channel then fill it with 0xff.
+    if (color_type == PNG_COLOR_TYPE_RGB || color_type == PNG_COLOR_TYPE_GRAY ||
+        color_type == PNG_COLOR_TYPE_PALETTE)
+      png_set_filler(png, 255, PNG_FILLER_AFTER);
+  }
+
+  // Since we read in BGR and write in BGR colors will not change
+  // + Comment out this condition and colors flip; read in RGB and write in BGR
+  if (colorspace == TILEDB_WEBP_BGR || colorspace == TILEDB_WEBP_BGRA) {
+    png_set_bgr(png);
+  }
+
+  if (color_type == PNG_COLOR_TYPE_GRAY ||
+      color_type == PNG_COLOR_TYPE_GRAY_ALPHA)
+    png_set_gray_to_rgb(png);
+
+  png_read_update_info(png, info);
+
+  // Set up buffers to hold rows of pixel data.
+  for (unsigned y = 0; y < *height; y++) {
+    size_t row_len = png_get_rowbytes(png, info);
+    auto row = (uint8_t*)(std::malloc(row_len));
```

Comment:
Are we supposed to use malloc here? Just a genuine question, since I feel as there is a standard that we should write code without malloc in the codebase.

Reply:
Unfortunately I don't believe there is a way to use libpng without malloc based on what I could find in the [libpng manual](http://www.libpng.org/pub/png/libpng-1.2.5-manual.html) or examples online. I'll do some more reading though, maybe I can figure a way around it by digging into the libpng API a bit more.